### PR TITLE
[Dev] docs(megatron-fsdp): add Megatron-FSDP user guide

### PR DESCRIPTION
**Pull request by xuwchen**
# What does this PR do ?
main PR: [#2396](https://github.com/NVIDIA/Megatron-LM/pull/2396) 

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by yanring** [comment]
#### `docs/discussions/megatron-fsdp-user-guide/example-scripts/sbatch_mfsdp_deepseek_v3.sh`

**Line 1:**
> 🔴 **Critical**: The shebang uses a full-width exclamation mark (`#！/bin/bash`) instead of ASCII (`#!/bin/bash`). This will prevent the script from executing properly. Please replace the character with the standard ASCII `!`.

---

#### `docs/discussions/megatron-fsdp-user-guide/megatron-fsdp-user-guide.md`

**Line 43:**
> Minor grammar suggestion: "This step avoids potential bubble**s** in **the** CUDA stream."

**Line 51:**
> Typo: "douhle buffers" → "double buffers"

**Line 67:**
> Typo: "registraion" → "registration"

**Line 76:**
> Typo: "registraion" → "registration"

**Line 88-89:**
> Consider adding a note clarifying where the output JSON file is saved:
> ```markdown
> This will create a `param_to_param_group_map.json` file in the `/path/to/param_to_param_group_map` directory.
> ```

---

#### `docs/discussions/megatron-fsdp-user-guide/example-scripts/sbatch_checkpoint_convert.sh`

**Line 50:**
> The file is missing a trailing newline. Please add a newline at the end for POSIX compliance.

**Comment by BestJuly** [comment]
LGTM. Could you please help delete the [Optimizing DeepSeek-V3 Training Performance on NVIDIA GB200 NVL72](https://github.com/NVIDIA/Megatron-LM/blob/dev/docs/discussions/deepseek-v3-gb200-optimization/deepseek-v3-gb200-optimization.md) part in the `README.md` in this PR because this has been deleted in commit [3c1b98e
](https://github.com/NVIDIA/Megatron-LM/commit/3c1b98efdc3add92460b9c78f6bd3aa5afebf58a)

**Review by yanring — APPROVED** [review]

File: docs/discussions/megatron-fsdp-user-guide/megatron-fsdp-user-guide.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+#### 2. Add `--calculate-per-token-loss`
+
+For gradients sharding mode optimization, include the `--calculate-per-token-loss` flag in your training script. This improves performance by reducing the frequency of gradient scaling, which is also a sizable drain on SM resources.
+
+#### 3. Add `--init-model-with-meta-device`
+
+Allows model initialization using meta device, followed by layer-by-layer initialization of distributed model weight buffers via the `Module.reset_parameters` API, facilitating the initialization of extremely large models.
+
+#### 4. Add `--grad-reduce-in-bf16`
```
Comment (Skylion007) [inline docs/discussions/megatron-fsdp-user-guide/megatron-fsdp-user-guide.md:55]: This has numerical issues at large scale though right? Like very large models at very large batch sizes? Or have you not observed this for pretraining very large 100B+ models?

File: docs/discussions/megatron-fsdp-user-guide/megatron-fsdp-user-guide.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+Enables gradient reduction in BF16 precision instead of FP32, reducing communication volume and accelerating the backward pass.
+
+#### 5. Add `--fsdp-double-buffer`
+
+Uses persistently allocated double buffers for temporarily-defined memory needed in `MegatronFSDP` communications. While having persistent double buffers may increase peak VRAM utilization, it is necessary to register NCCL user buffers (`nccl_ub=True`) for `MegatronFSDP`. Currently, this is supported only for simple repetitive model structures such as GPT.
+
+- **Only effective when using Megatron-LM.**
+- Defaults to `False`. Automatically overridden to `True` when `nccl_ub` is enabled.
+
+#### 6. Add `--use-nccl-ub`
```
Comment (Skylion007) [inline docs/discussions/megatron-fsdp-user-guide/megatron-fsdp-user-guide.md:66]: May want to add a note this doesn't work with segmentable allocator?

**Comment by BoxiangW** [comment]
Thanks Xuwen for this great documentation, can you also move it or at least link it [here ](https://docs.nvidia.com/megatron-core/developer-guide/latest/api-guide/custom_fsdp.html) ([code](https://github.com/NVIDIA/Megatron-LM/blob/main/docs/source/api-guide/custom_fsdp.md))? Not sure where is this doc/discussion being published.

File: docs/discussions/megatron-fsdp-user-guide/megatron-fsdp-user-guide.md (hunk 2 above)
Comment (xuwchen) [reply docs/discussions/megatron-fsdp-user-guide/megatron-fsdp-user-guide.md:66]: You're right. I've added a note about the incompatibility with the segmentable allocator (PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True). The code already performs a [runtime check](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/distributed/distributed_data_parallel_config.py#L153) and raises an error when both are enabled, but it's better to document this upfront in the user guide.

File: docs/discussions/megatron-fsdp-user-guide/megatron-fsdp-user-guide.md (hunk 1 above)
Comment (xuwchen) [reply docs/discussions/megatron-fsdp-user-guide/megatron-fsdp-user-guide.md:55]: Thanks for bringing this up. Have you run into concrete numerical issues at very large model and batch sizes in your experiments?

File: docs/discussions/megatron-fsdp-user-guide/megatron-fsdp-user-guide.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+- **Only effective when using Megatron-LM.**
+- Defaults to `False`.
+- By default we try to use NCCL window (symmetric) registration if it is available. If not it falls back to conventional local registration.
+- **Incompatible with PyTorch's segmentable allocator:** Do not set `PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True` when using `--use-nccl-ub`, as this will cause a runtime error due to compatibility issues with the `torch.cuda.MemPool` API.
+
+## Checkpoint Conversion from 3D-Parallel to Megatron-FSDP
+
+Megatron-FSDP introduces a new checkpoint format `fsdp_dtensor`. To help you smoothly transition from 3D-Parallel to Megatron-FSDP, we provide a script for converting checkpoints from the `torch_dist` format to the `fsdp_dtensor` format. Using DeepSeek-V3 as an example, the detailed conversion process is described below.
```
Comment (zhujian19891203) [inline docs/discussions/megatron-fsdp-user-guide/megatron-fsdp-user-guide.md]: Thanks for your contribution, it's really a great starting point for me to try the Megatron-FSDP. Since our subsequent evaluation tasks are based on the torch_dist ckpt type, could you provide a path to convert ckpt format from fsdp_dtensor to torch_dist or huggingface? Or huggingface format?

Comment (xuwchen) [reply docs/discussions/megatron-fsdp-user-guide/megatron-fsdp-user-guide.md]: For checkpoint conversion, this feature will be added to Megatron-Bridge and is currently under development. Once it is ready, it will support converting between fsdp_dtensor and Hugging Face formats. Since Megatron-Bridge already supports conversion between torch_dist and Hugging Face formats, this will make it straightforward to go from fsdp_dtensor to torch_dist as well.
